augur/merge.py

```diff
@@ -354,6 +379,57 @@ def merge_metadata(args):
             print_info(f"WARNING: Skipped deletion of {db_path} due to error, but you may want to clean it up yourself (e.g. if it's large).")
 
 
+def merge_sequences(args):
+    if not args.skip_input_sequences_validation:
+        for s in args.sequences:
+            print_info(f"Validating {s!r}…")
+
+            with NamedTemporaryFile() as dup_num_file:
+                command = f"""
+                    {augur} read-file {shquote(s)} |
+                    {seqkit()} rmdup --dup-num-file {shquote(dup_num_file.name)}
+                """
+
+                returncode = run_command(command, print_stdout=False)
+
+                if returncode != 0:
+                    raise AugurError(f"Validation failed for {s!r}.")
+
+                dup_num_file.seek(0)
+                
+                if duplicates := list(parse_seqkit_dup_num_file(dup_num_file.read().decode().strip())):
+                    raise AugurError(f"File {s} has duplicate IDs: {', '.join(repr(d) for d in duplicates)}")
+
+    print_info(f"Merging sequences and writing to {args.output_sequences!r}…")
+
+    # Reversed because seqkit rmdup keeps the first entry but this command
+    # should keep the last entry.
+    # "echo;" adds a newline character to support FASTA files that are missing one at the end.
+    # This is fine because extraneous newline characters are stripped by seqkit.
+    command = f"""
+        (for f in {" ".join(shquote(s) for s in reversed(args.sequences))}; do
+            {augur} read-file "$f";
+            echo;
+        done) |
+        {seqkit()} rmdup |
+        {augur} write-file {shquote(args.output_sequences)}
+    """
+
+    returncode = run_command(command, print_stdout=True)
+
+    if returncode != 0:
+        # Ideally, read-file, seqkit, and write-file errors would all be handled
+        # separately. That is not possible because everything is done in one
+        # child process with one return code.
+        # FIXME: also clean up metadata merge files, and vice versa?
```

Comment:
clean them up under what conditions?

Reply:
If both metadata and sequence merging are requested, and sequence merge fails, we'll clean up any partial sequence merge outputs. Should metadata merge outputs be cleaned up as well? Based on the current sequential order of metadata merge followed by sequence merge, which is somewhat arbitrary, metadata outputs would be complete upon sequence merge failure, but maybe they should be removed along with the partial sequence merge outputs.

Reply:
+1 for "leave the metadata outputs but be real loud about sequence failure meaning maybe they're wrong", if that's an option that's not too onerous to implement…

Reply:
Added the error handling at the top level:

https://github.com/nextstrain/augur/blob/1979c61036850e5de021908797b7f15be06267d0/augur/merge.py#L151-L159

The output looks like this:

https://github.com/nextstrain/augur/blob/1979c61036850e5de021908797b7f15be06267d0/tests/functional/merge/cram/merge-sequences-errors.t#L24-L36